crates/supervisor/core/src/safety_checker/task.rs

```diff
@@ -0,0 +1,255 @@
+                            debug!(
+                                target: "safety_checker",
+                                chain_id = self.chain_id,
+                                target_level = %self.target_level,
+                                %block_info,
+                                "Promoted next candidate block"
+                            );
+                        }
+                        Err(err) => {
+                            warn!(
+                                target: "safety_checker",
+                                chain_id = self.chain_id,
+                                target_level = %self.target_level,
+                                %err,
+                                "Error promoting next candidate block"
+                            );
+                            tokio::time::sleep(self.interval).await;
+                        }
+                    }
+                } => {}
+            }
+        }
+
+        info!(target: "safety_checker", chain_id = self.chain_id, target_level = %self.target_level, "Stopped safety checker");
+    }
+
+    // Attempts to promote the next block at the current safety level,
+    // after validating cross-chain dependencies.
+    fn promote_next_block(
+        &self,
+        checker: &CrossSafetyChecker<'_, P>,
+    ) -> Result<BlockInfo, CrossSafetyError> {
+        let candidate = self.find_next_promotable_block()?;
+
+        checker.verify_block_dependencies(self.chain_id, candidate, self.target_level)?;
+
+        // TODO: Add more checks in future
+
+        self.provider.update_safety_head_ref(self.chain_id, self.target_level, &candidate)?;
+
+        Ok(candidate)
+    }
+
+    // Finds the next block that is eligible for promotion at the current target level.
+    fn find_next_promotable_block(&self) -> Result<BlockInfo, CrossSafetyError> {
+        let promotion_boundary = self.promotion_upper_bound()?;
+
+        let current_head = self.provider.get_safety_head_ref(self.chain_id, self.target_level)?;
+        let upper_head = self.provider.get_safety_head_ref(self.chain_id, promotion_boundary)?;
+
+        if current_head.number >= upper_head.number {
+            return Err(CrossSafetyError::NoBlockToPromote);
+        }
+
+        let candidate = self.provider.get_block(self.chain_id, current_head.number + 1)?;
+
+        Ok(candidate)
+    }
+
+    // Returns the safety level that defines the upper promotion boundary for the current level.
+    //
+    // For example:
+    // - CrossUnsafe promotions are bounded by LocalUnsafe.
+    // - CrossSafe promotions are bounded by LocalSafe.
+    const fn promotion_upper_bound(&self) -> Result<SafetyLevel, CrossSafetyError> {
+        match self.target_level {
+            SafetyLevel::CrossUnsafe => Ok(SafetyLevel::LocalUnsafe),
+            SafetyLevel::CrossSafe => Ok(SafetyLevel::LocalSafe),
+            _ => Err(CrossSafetyError::UnsupportedTargetLevel(self.target_level)),
+        }
+    }
+}
```

Comment:
the target level doesn't change as the program runs, so this can be done once as part of the constructor, and stored in a new files of the type `promotion_upper_bound`